lib/semantics/check-omp-structure.cc

```diff
@@ -48,23 +48,63 @@ bool OmpStructureChecker::HasInvalidWorksharingNesting(
 
 void OmpStructureChecker::CheckAllowed(OmpClause type) {
   if (!GetContext().allowedClauses.test(type) &&
-      !GetContext().allowedOnceClauses.test(type)) {
+      !GetContext().allowedOnceClauses.test(type) &&
+      !GetContext().allowedExclusiveClauses.test(type)) {
     context_.Say(GetContext().clauseSource,
         "%s clause is not allowed on the %s directive"_err_en_US,
         EnumToString(type),
         parser::ToUpperCaseLetters(GetContext().directiveSource.ToString()));
     return;
   }
-  if (GetContext().allowedOnceClauses.test(type) && FindClause(type)) {
+  if ((GetContext().allowedOnceClauses.test(type) ||
+          GetContext().allowedExclusiveClauses.test(type)) &&
+      FindClause(type)) {
     context_.Say(GetContext().clauseSource,
         "At most one %s clause can appear on the %s directive"_err_en_US,
         EnumToString(type),
         parser::ToUpperCaseLetters(GetContext().directiveSource.ToString()));
     return;
   }
+  if (GetContext().allowedExclusiveClauses.test(type)) {
+    std::vector<OmpClause> others;
+    GetContext().allowedExclusiveClauses.IterateOverMembers([&](OmpClause o) {
+      if (FindClause(o)) others.emplace_back(o);
+    });
+    if (!others.empty()) {
+      context_.Say(GetContext().clauseSource,
+          "%s and %s are mutually exclusive and may not appear on the same %s directive"_err_en_US,
+          EnumToString(type), EnumToString(others[0]),
+          parser::ToUpperCaseLetters(GetContext().directiveSource.ToString()));
+      return;
+    }
+  }
   SetContextClauseInfo(type);
 }
 
+void OmpStructureChecker::RequiresConstantPositiveParameter(
+    const OmpClause &clause, const parser::ScalarIntConstantExpr &i) {
```

Comment:
Ah yes I should have mentioned that, I didn't apply this in the ORDERED clause because for some reason the integer contained in the Ordered clause node is a ScalarIntExpr instead of a ScalarIntConstantExpr, even though semantically it can only be a constant. I'm not familiar with the parser but can probably dig around to work out why that is, I just thought that might be better done as a separate patch.

Reply:
I saw that but for some reason when I do x.v on the Ordered struct that's passed into this function I get a ScalarIntExpr, which is why I wanted to dig into this more. I'll take a look tomorrow to see if I can work out why that is

Reply:
Ah, yes, my mistake it's the fact that it's wrapping and optional and I was trying to pass x.v instead of *expr (aka *x.v).